Scheduledjob e2e

soltysh opened the pull request:
@erictune last element of the scheduledjob puzzle. I think we'll iterate on this once we have all the puzzles in place. This is one of those things that will be allowed to merge after code freeze.

``` release-note
* Introducing ScheduledJobs as described in [the proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/scheduledjob.md) as part of `batch/v2alpha1` version (experimental feature).
```

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/26027)

<!-- Reviewable:end -->

soltysh committed:
ScheduledJob e2e

erictune commented on test/e2e/scheduledjob.go:
```diff
@@ -0,0 +1,178 @@
+		err = waitForActiveJobs(f.Client, f.Namespace.Name, scheduledJob.Name, 2)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Removing scheduledjob")
```
Should deleting a scheduledJob delete its jobs?

soltysh replied:
Hmmm... that's something I wasn't sure about. With current approach I've decided that at least removing the SJ will not schedule more jobs, especially that the schedule is pretty dense (new job every 5 sec.). The question is do we care about the jobs after the SJ is removed? I'm inclined to say no. Lemme check how deployments treat older RS, if removing those will also remove them.

soltysh replied:
From my quick skim over the `DeploymentReaper` it looks like it removes all related `ReplicaSet`s. In that case I'd say yes remove all.

soltysh commented:
This is blocked by #25952.

janetkuo commented:
Running all three e2e tests in #29137 gives me
```
   Expected error:
      <*runtime.notRegisteredErr | 0xc82046c680>: {
          gvk: {Group: "batch", Version: "v1", Kind: "ScheduledJob"},
          t: nil,
      }
      no kind "ScheduledJob" is registered for version "batch/v1"
  not to have occurred
```

soltysh commented:
I haven't got a chance to verify this properly at the time of writing. I'll
revisit this while looking at the remaining PRs we have for SJ.

On Aug 4, 2016 8:02 PM, "Janet Kuo" <notifications@github.com> wrote:

> Running all three e2e tests in #29137
> <https://github.com/kubernetes/kubernetes/pull/29137> gives me
>
>    Expected error:
>       <*runtime.notRegisteredErr | 0xc82046c680>: {
>           gvk: {Group: "batch", Version: "v1", Kind: "ScheduledJob"},
>           t: nil,
>       }
>       no kind "ScheduledJob" is registered for version "batch/v1"
>   not to have occurred
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26027#issuecomment-237633620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjLVWylqkGMpxFZUnQtt3k9USTG82Ucks5qcik6gaJpZM4Ij2-x>
> .
>

soltysh committed:
Allow setting GroupVersion in e2e tests

soltysh commented:
This relies on https://github.com/kubernetes/kubernetes/pull/30227, I've added option to set the testing group, so no more `no kind "ScheduledJob" is registered for version "batch/v1"` error :wink: 
The only problem I have right now is the actual tests, but I'll be working on that and the fact that this group is disabled by default, should I just force enabling everything for e2e, @janetkuo you know by any chance?

janetkuo commented:
Yes we can enable `batch/v2alpha1` in e2e test, but do we also want to test all e2e tests work fine without enabling batch/v2alpha1? @lavalamp may know?

janetkuo commented on test/e2e/scheduledjob.go:
```diff
@@ -0,0 +1,184 @@
+)
+
+var _ = framework.KubeDescribe("ScheduledJob", func() {
+	options := framework.FrameworkOptions{
```
Or can we do something like:

```go
f := framework.NewDefaultFramework("scheduledjob")
f.GroupVersion = &unversioned.GroupVersion{Group: batch.GroupName, Version: "v2alpha1"}
```

soltysh replied:
I'd propose leave this as is for now, the end goal is to entirely get rid of the unversioned client, then the problem should disappear and SJ is first on my list, but that will be not sooner than 1.5.

janetkuo commented on test/e2e/scheduledjob.go:
```diff
@@ -0,0 +1,184 @@
+
+const (
+	// How long to wait for a scheduledjob
+	scheduledJobTimeout = 15 * time.Minute
```
Can we have a shorter timeout? Is 5 min enough?

janetkuo committed:
Use unversioned client in scheduledjobs and set group version to batch/v2alpha1

janetkuo commented:
I found a bug in sj controller and you'll need this PR #30327

soltysh commented:
@kubernetes/sig-testing how can I globally enable batch/v2alpha1 for e2e tests?

soltysh committed:
Enable batch/v2alpha1 globally for e2e tests

soltysh commented:
Rebased and addressed comments, @janetkuo I've picked your PR for now ptal the current changes.
@lavalamp || @kubernetes/sig-testing ptal the last commit enabling `batch/v2alpha1` globally in e2e tests

janetkuo commented:
A couple of thoughts / suggestions:

1. Do we need to test that no jobs are running after sj is deleted? (Do we have the reaper already?)
2. List jobs (in addition to check sj status) when verifying we have correct number of jobs
3. Test ReplaceConcurrent (if this takes too much time, can be a follow-up PR)

janetkuo commented:
Reviewed 6 of 9 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027)*
<!-- Sent from Reviewable.io -->

spiffxp commented on hack/e2e.go:
```diff
@@ -262,6 +262,8 @@ func TearDown() bool {
 // Up brings an e2e cluster up, recreating it if one is already running.
 func Up() bool {
+	// force having batch/v2alpha1 always on for e2e tests
+	os.Setenv("KUBE_RUNTIME_CONFIG", "batch/v2alpha1=true")
```
I get that this is DRY'er than setting a default in all of the `clusters/*/config-foo.sh` scripts but I don't know if it's a pattern we want to extend as more features get added?

soltysh replied:
Right, that's why I asked. Again with `batch/v2alpha1` I'm blazing the trail, this time for group versions for e2e tests.

soltysh commented:
> Do we need to test that no jobs are running after sj is deleted? (Do we have the reaper already?)

No reaper yet, need to write one. Will work on it tomorrow.

> List jobs (in addition to check sj status) when verifying we have correct number of jobs

Done.

> Test ReplaceConcurrent (if this takes too much time, can be a follow-up PR)

Done, but I aslo found a bug there, see https://github.com/kubernetes/kubernetes/issues/30442

I'll rebase this once your PR lands.

janetkuo commented:
Reading #30442 makes me think that we should verify pods created by jobs as well. 

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[test/e2e/scheduledjob.go, line 60 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027#-KOvNcC3ayRaEUJkFIBv:-KOvNcC3ayRaEUJkFIBw:bftv9xg) ([raw file](https://github.com/kubernetes/kubernetes/blob/c0eded387a20b6052c75d3089c4d65f14d162916/test/e2e/scheduledjob.go#L60)):*
> ```Go
> 		jobs, err := f.Client.Batch().Jobs(f.Namespace.Name).List(api.ListOptions{})
> 		Expect(err).NotTo(HaveOccurred())
> 		Expect(jobs.Items).To(HaveLen(2))
> ```

Could there be more than two jobs?

---

*[test/e2e/scheduledjob.go, line 233 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027#-KOvRJzhdRrG0qx7kfaj:-KOvRJzhdRrG0qx7kfak:b-yp4cj9) ([raw file](https://github.com/kubernetes/kubernetes/blob/c0eded387a20b6052c75d3089c4d65f14d162916/test/e2e/scheduledjob.go#L233)):*
> ```Go
> 
> // Wait for a new job to appear.
> func waitForNewJob(c *client.Client, ns, previousJobName string) error {
> ```

Optional: `waitForNewJob` may be misused by putting expected job name as the last parameter. How about `waitForJobReplaced`?

---

*[test/e2e/scheduledjob.go, line 240 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027#-KOvQabY7F4UTjNeczXC:-KOvQabY7F4UTjNeczXD:b-4bjadt) ([raw file](https://github.com/kubernetes/kubernetes/blob/c0eded387a20b6052c75d3089c4d65f14d162916/test/e2e/scheduledjob.go#L240)):*
> ```Go
> 		}
> 		if len(jobs.Items) != 1 {
> 			return false, nil
> ```

If there are more than 1 jobs, should we return error?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027#-:-KOvS9fwO4zEs072BhwI:b4et598)*
<!-- Sent from Reviewable.io -->

soltysh commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[test/e2e/scheduledjob.go, line 60 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027#-KOvNcC3ayRaEUJkFIBv:-KOzEvqBqaInLpPCRwRG:baxwyre) ([raw file](https://github.com/kubernetes/kubernetes/blob/c0eded387a20b6052c75d3089c4d65f14d162916/test/e2e/scheduledjob.go#L60)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Could there be more than two jobs?
</details>
Yeah, it can, I'll change this check to be >= 2

---

*[test/e2e/scheduledjob.go, line 233 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027#-KOvRJzhdRrG0qx7kfaj:-KOzF36HHQcIOQod7Mns:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/c0eded387a20b6052c75d3089c4d65f14d162916/test/e2e/scheduledjob.go#L233)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Optional: `waitForNewJob` may be misused by putting expected job name as the last parameter. How about `waitForJobReplaced`?
</details>
Done

---

*[test/e2e/scheduledjob.go, line 240 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027#-KOvQabY7F4UTjNeczXC:-KOzF4fOhOCAF4E4wKFv:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/c0eded387a20b6052c75d3089c4d65f14d162916/test/e2e/scheduledjob.go#L240)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> If there are more than 1 jobs, should we return error?
</details>
Done

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027)*
<!-- Sent from Reviewable.io -->

soltysh commented:
> Reading #30442 makes me think that we should verify pods created by jobs as well.

If you don't mind I'll leave it as a follow up because of that particular issue.

janetkuo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/26027#-:-KP-2syrokgBDbUGJ_fL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

soltysh commented:
I'm removing lgtm label, because this needs #30327 first

soltysh commented:
Apparently it was too late.

girishkalele commented:
@soltysh The submit queue is blocked since 13:40 due to GKE failures in the ScheduledJob e2e test.
Is the other commit #30327 going to go in soon ?
Otherwise, we should revert this commit till the other goes in ?

janetkuo commented:
Failure on GKE:
```
[k8s.io] ScheduledJob
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:714
  should not schedule new jobs when ForbidConcurrent [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/scheduledjob.go:118

  Expected error:
      <*errors.StatusError | 0xc820bfc300>: {
          ErrStatus: {
              TypeMeta: {Kind: "Status", APIVersion: "v1"},
              ListMeta: {SelfLink: "", ResourceVersion: ""},
              Status: "Failure",
              Message: "the server could not find the requested resource",
              Reason: "NotFound",
              Details: {Name: "", Group: "", Kind: "", Causes: nil, RetryAfterSeconds: 0},
              Code: 404,
          },
      }
      the server could not find the requested resource
  not to have occurred
```

janetkuo commented:
Does `KUBE_RUNTIME_CONFIG` work on GKE cluster? (Trying to figure out why it failed on GKE)

janetkuo commented:
Maybe set `RUNTIME_CONFIG` instead?

soltysh commented:
> Maybe set RUNTIME_CONFIG instead?

The pattern I've noticed is that I set `KUBE_RUNTIME_CONFIG` and the script uses that to set `RUNTIME_CONFIG` because it might want to add something more to that env. I'll look into GKE scripts and update those accordingly.
